XBeeModuleSim.py

```diff
@@ -0,0 +1,197 @@
+
+
+class XBeeModuleSim:
+    # Each element in each queue is a bytearray.
+    _rocket_rx_queue_packed = SimpleQueue()  # RX from RKT
+    _ground_rx_queue_packed = SimpleQueue()  # RX from GND
+
+    def send_to_rocket(self, data):
+        """
+        :brief: Queue data to send to rocket following the XBee protocol.
+        :param data: bytearray of data.
+        """
+        self._ground_rx_queue_packed.put(data)
+
+    def recieved_from_rocket(self, data):
+        """
+        :brief: All data incoming from the rocket should be passed into this method for processing.
+        :param data: bytearray of data recieved.
+        """
+        self._rocket_rx_queue_packed.put(data)
+
+    def __init__(self):
+        """
+        Constructor.
+        In addition to constructing this, if any useful work is to be done then the rocket_callback and ground_callback attributes should be set - by default they are simply no-op functions.
+        """
+        self._rocket_rx_queue = self._unpack(self._rocket_rx_queue_packed)
+
+        # Callbacks should be of form callback(data), where data is a bytearray.
+        # Making a no-op callback to prevent need for checking if callback exists.
+        def nop_callback(data):
+            return
+
+        """Is called whenever data needs to be sent to the rocket through SIM - this should be the SIM send() function. The callback should be thread safe."""
+        self.rocket_callback = nop_callback
+
+        """Is called whenever data recieved from the rocket needs to be sent to the rest of the ground station code. The callback should be thread safe."""
+        self.ground_callback = nop_callback
+
+        self._running_lock = Lock()
+        self._running = True
+
+        # Queues are IO bound.
+        self._rocket_rx_thread = Thread(
+            target=self._run_rocket_rx, name="xbee_sim_rocket_rx"
+        )
+        self._ground_rx_thread = Thread(
+            target=self._run_ground_rx, name="xbee_sim_ground_rx"
+        )
+        self._rocket_rx_thread.start()
+        self._ground_rx_thread.start()
+
+    def shutdown(self):
+        """ 
+        Shuts down everything and kills the internal threads.
+        Callbacks will not be interrupted/passed corrupt data.
+        """
+        with self._running_lock:
+            self._running = False
+
+        # Fill up the queues with garbage to unblock the threads on input
+        self._ground_rx_queue_packed.put(b"\0")
+        self._rocket_rx_queue_packed.put(b"\0")
+
+        self._rocket_rx_thread.join()
+        self._ground_rx_thread.join()
+
+    @property
+    def running(self):
+        with self._running_lock:
+            return self._running
+
+    def _unpack(self, q):
+        """
+        Helper generator that unpacks the iterables in a given queue.
+        """
+        while self.running:
+            arr = q.get()
+            for i in arr:
+                yield i
+        sys.exit()
```

Comment:
Are you sure? Where does it say that? https://docs.python.org/3/library/sys.html says:
"Exit from Python. This is implemented by raising the SystemExit exception, so cleanup actions specified by finally clauses of try statements are honored, and it is possible to intercept the exit attempt at an outer level."

I assume by "Python" they mean the Python process, which is a parent of all the threads.

Reply:
`sys.exit()` raises an (uncaught) exception, and only the thread terminates on an uncaught exception (unless you specifically specify when creating the thread to propagate exceptions)
Alternative is either create a new custom exception (probably overkill) or find some other built in exception

Reply:
Huh, I always thought python would terminate the entire process when there's an uncaught exception in a thread, like in C++ and other languages. I guess not. Maybe just add a comment explaining it's purpose and how it works. It could be a little confusing for anyone looking at this for the first time as to what `_unpack()` has to do with exiting.